SUSE brand toggle

mantis-toboggan-md opened the pull request:
#2506

mantis-toboggan-md commented:
Updated to
* remove brand settings page
* remove reference to customizing colors at runtime (finishing support for this tracked in #2794 )
* add brand to advanced settings view (when relevant; the setting wont appear until a support key has been provided)

lvuch reviewed (commented):
This is looking good. I'm removing the link variable as we will keep all links/outlines blue.

vincent99 commented on mixins/brand.js:
```diff
@@ -0,0 +1,67 @@
+  },
+
+  watch: {
+    brandSetting(neu) {
+      const uiPLSetting = findBy(this.globalSettings, { id: 'ui-pl' });
+      const brand = neu.value;
+
+      try {
+        const brandMeta = require(`~/assets/brand/${ brand }/metadata.json`);
+        const uiPL = brandMeta['ui-pl'] || uiPLSetting.default;
+
+        uiPLSetting.value = uiPL;
+      } catch {
+        uiPLSetting.value = uiPLSetting.default;
+      }
+
+      uiPLSetting.save();
+    },
+  },
```
It's also currently an infinite loop of death...

![image](https://user-images.githubusercontent.com/753917/116486790-4511c600-a843-11eb-9e01-87d3974296f3.png)

vincent99 commented on store/prefs.js:
```diff
@@ -414,4 +419,37 @@ export const actions = {
+      try {
+        const brandSetting = rootGetters['management/byId'](MANAGEMENT.SETTING, 'brand');
+
+        if (brandSetting && brandSetting.value && brandSetting.value !== '') {
+          const brand = brandSetting.value;
```
@nwmac @mantis-toboggan-md Do we _need_ this for 2.6?  The idea of having alternate styles for rancher vs suse is fine, but I just do not like anything about this custom theme implementation and would like to just drop it.

- I have to get a file into assets/brand at build-time somehow; how does this help a user actually apply their own theme at runtime?
- It's in an uncanny-valley that allows users to override only a small part of the theme.  Most customers are either fine with the theme as it is, or are going to want to customize more than just one color.  E.g. some color choices will require changing other colors for text on top of it to be readable; the "what contrasts more" function is not really a good solution for that.
- It's done in JS, which means it doesn't work in SSR.  You see the regular theme and then the brand value seconds later after rehydration is done.
- It's duplicating all sorts of chunks of the SCSS rules and functions with more JS that is brittle and has to be kept in sync
- It's breaking things if used that aren't broken with the hasStylesheet=true version, like the login page: 
![image](https://user-images.githubusercontent.com/753917/116487824-ba7e9600-a845-11eb-8942-957d244a574c.png)

vincent99 commented on assets/styles/themes/_suse_dark.scss:
```diff
@@ -0,0 +1,4 @@
+.suse-dark {
+    @include theme-dark($primary: rgb(48, 186, 120));
```
I don't understand why the whole mixin and defining/duplicating/loading every possible rule of the theme is needed.  The idea of the themeing is/was:

- Everything theme-able is exposed as a CSS variable
- Related things should refer to an existing variable so they only need to be changed once unless you really want them to be different (e.g. `--thing-border: var(--border)`)
- The "root" theme (which is `light`) is always loaded (and is selected by `::root`)
- Other themes overwrite just what should _differ_ from the root theme (and are selected by `.theme-$name` applied to the body)
- Zero or one of those is selected and applied to the body
- Everything works via plain rendering & CSS selectors, without involving JS, so that it renders right the first time from SSR HTML.

Because of laziness/convenience there is a fair bit of scss-computed calculating colors going on in there now, but rather than doubling down on that with JS implementations we should be getting rid of those by having things that are the primary color use `var(--primary)` instead of `$primary`, etc.

The main use case you're interested in here of changing just the primary color should ideally consist of:

```
.theme-suse-light {
  --primary: rgb(48,186,120);
 ... maybe a few more ...
}
```

And everything else would extend from that.

vincent99 reviewed (changes requested):
Besides comments inline, a few other things...

- This doesn't work on the login page.  Login saves the theme into a cookie so that it knows which one is (probably) supposed to be used and you don't see it change after login.

- This breaks the 'auto' theme switching based on the OS preferred-color-scheme.

mantis-toboggan-md commented:
Updates:

* removed editing of ui-pl
* single suse stylesheet applied in addition to dark and light themes
* always apply light theme
    - I found that when I used `:root` I was unable to override the css variables calculated from `--primary` merely by overriding `--primary` in the suse theme; using `BODY` instead fixes this 
* use css vars in dark and light themes as much as possible
* pull `primary-` and `info-` color class generation out of the `color-variables` mixin to simplify the suse stylesheet (maybe not important now but helpful for white labeling later)
* set brand setting cookie to attempt to load brand theming on login page
    -  if we want branding applied on unauthenticated pages we need a publicly-visible global setting; I know some of those already exist eg `first-login` but unsure how a new one would be added

lvuch reviewed (approved):
lgtm

nwmac commented:
@mantis-toboggan-md Looks great (I've not gone through the code yet) - a few things I spotted:

- The logo on the support/home page does not update straight away when you enter a subscription ID
- The 'glass' background color used for loading and for dialogs is green for the SUSE theme - which looks jarring - I think we can keep it as before
- The text color in the top-level menu for hover state needs to be white for light theme:
![image](https://user-images.githubusercontent.com/1955897/117664664-f1d03980-b199-11eb-9c84-a9f500e3a72b.png)

- The SUSE Rancher logo doesb't look great! @lvuch Can you work any magic?

Other suggestions:

- Instead of having the input box on the support page, can we have that in a popup dialog and then have a button on the page to show - this would look cleaner, I think
- Related to above - can we provide a way to remove the subscription ID, so it reverts back fo Rancher - perhaps we use the same dialog, but if registered, doesn't take an input, but when confirmed, removes the ID

lvuch commented:
@nwmac @mantis-toboggan-md colors.scss line 9 `mixin` -- thoughts on rewriting this?

mantis-toboggan-md commented:
I've implemented most of your suggestions @nwmac with a couple exceptions:

> * The text color in the top-level menu for hover state needs to be white for light theme:

I fixed the hover state color but thought it still looked weird, asked @lvuch offline she suggested this gray instead:
![Screen Shot 2021-05-11 at 2 04 43 PM](https://user-images.githubusercontent.com/42977925/117891217-5f986600-b26b-11eb-85b1-227d1ff95916.png)


> * Can we update the welcome message on the home screen to use the product name - I think we said if rancher-pl was not set, we would use 'Rancher' if not supported, 'SUSE Rancher' if supported, otherwise the value of rancher-pl - would be good to have a common fn for that and be able to feed that on the home page, so it says "Welcome to SUSE Rancher"

I'm going to add this to [this PR](https://github.com/rancher/dashboard/pull/2944) instead since there's ui-pl work already going on there.

nwmac commented:
@mantis-toboggan-md he colors still aren't right.

I'm not sure about the witch to HSL.

Can we take the color changes out of this PR and create new SUSE sheets with the color changes to primary.

Then we can get this in and work out how to do the color changes - i think we need to rework the light and dark stylesheets - some of what you've already done here, but we can discuss how to remove the color mixin that is used.

mantis-toboggan-md commented:
> Can we take the color changes out of this PR and create new SUSE sheets with the color changes to primary.

Just to clarify, are you asking that I revert changes in `_light.scss` and `_dark.scss` and make the suse scss file handle all the varaibles calculated from `$primary`? Or are you asking that I remove color-changing functionality from this PR altogether?

mantis-toboggan-md commented:
Updates:
* revert `--overlay-bg` (loading background) to light grey
* use `--primary-hover-bg` and `--primary-hover-text` in `TopLevelNav`
* revert HSL changes to `_light.scss `and `_dark.scss` and define more vars in `suse.scss` instead
    *  @lvuch updated the dropdowns to stay blue in this commit https://github.com/rancher/dashboard/pull/2788/commits/e24a44345e859014f31e6923bbf030d6255bd096 and I've preserved that here, though I did accidentally remove it in a previous commit.

nwmac reviewed (approved):
Thanks @mantis-toboggan-md for reverting the custom theming changes - this looks good to me - there is a minor issue in the dialog that lets you enter the subscription ID - when you submit, it flips to subscribed/unsubscribed just before closing, as we update the setting, that gets picked up before the change is made - I suggest we get this in as is and address that if we want to later.

Would be great to get this in today

nwmac commented:
@mantis-toboggan-md This needs a rebase!
